### PR TITLE
Cuda matrix conversion sellp

### PR DESCRIPTION
**Pull request by fritzgoebel**
Conversion kernels between dense and sellp.

Requirements:
+ [x] Merge PR #198

### CODE REVIEW

**Comment by tcojean** [comment]
To see and review the changes of this MR so far, you can use this link:
https://github.com/ginkgo-project/ginkgo/pull/224/files/342e5b184d503b87ee5bf8589ff8279d74e1e9a5..2de3d123ed91d2ab2a8936749ac35b7d45b981d5

**Comment by tcojean** [comment]
@fritzgoebel on CUDA 10.0 (and only there), the dense_kernels test fail, both with clang and gcc:
https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/160720369
https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/160720371

File: cuda/matrix/sellp_kernels.cu [hunk 1]
```diff
@@ -160,10 +162,80 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+                          cuda_config::max_block_size / cuda_config::warp_size,
+                          1);
+    const dim3 init_grid_dim(ceildiv(result->get_stride(), block_size.x),
+                             ceildiv(num_rows, block_size.y), 1);
+
+    kernel::initialize_zero_dense<<<init_grid_dim, block_size>>>(
+        num_rows, num_cols, result->get_stride(),
+        as_cuda_type(result->get_values()));
+
+    const dim3 block_dim(slice_size, default_block_size / slice_size, 1);
```
Comment (thoasm) [inline cuda/matrix/sellp_kernels.cu]: What if `default_block_size < slice_size`, or if `default_block_size % slice_size != 0`? Either way, the result will likely be wrong.

I would recommend using
```c++
constexpr int threads_per_row = cuda_config::warp_size; // if you want multiple threads per row
const dim3 block_dim(default_block_size, 1, 1);
const dim3 grid_dim(ceildiv(slice_size * slice_num * threads_per_row, default_block_size), 1, 1);
```
(in combination with the proposed changes above).

File: cuda/test/matrix/dense_kernels.cpp [hunk 2]
```diff
@@ -98,10 +106,10 @@ class Dense : public ::testing::Test {
 
     void set_up_apply_data()
     {
-        x = gen_mtx<Mtx>(40, 25);
-        c_x = gen_mtx<ComplexMtx>(40, 25);
+        x = gen_mtx<Mtx>(65, 25);
+        c_x = gen_mtx<ComplexMtx>(65, 25);
         y = gen_mtx<Mtx>(25, 35);
-        expected = gen_mtx<Mtx>(40, 35);
+        expected = gen_mtx<Mtx>(65, 35);
```
Comment (thoasm) [inline cuda/test/matrix/dense_kernels.cpp]: Out of curiosity, why did you change these numbers? I am fine with the change, I would like to know why you did it.

Comment (fritzgoebel) [reply cuda/test/matrix/dense_kernels.cpp]: I did it because the default slice size is 64 and I wanted to test with at least two slices.

Comment (thoasm) [reply cuda/test/matrix/dense_kernels.cpp]: Makes sense.

File: cuda/matrix/dense_kernels.cu [hunk 3]
```diff
@@ -686,11 +687,134 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+        size_type *__restrict__ slice_lengths,
+        size_type *__restrict__ slice_sets)
+{
+    constexpr auto warp_size = cuda_config::warp_size;
+    const auto sliceid = blockIdx.x;
+    const auto tid_in_warp = threadIdx.x;
+
+    if (sliceid * slice_size + tid_in_warp < num_rows) {
+        size_type thread_result = 0;
+        for (auto i = tid_in_warp; i < slice_size; i += warp_size) {
```
Comment (thoasm) [inline cuda/matrix/dense_kernels.cu]:
```suggestion
        for (size_type i = tid_in_warp; i < slice_size; i += warp_size) {
```
